core/arch/x86/encode.c

```diff
@@ -1053,6 +1041,12 @@ reg_size_ok(decode_info_t *di /*prefixes field is IN/OUT; x86_mode is IN*/, reg_
         }
         return true;
     }
+    /* XXX i#1312: we haven't really resolved the register's size. The mask register
+     * size is dependent on the machine's support for certain AVX-512 derivatives. We may
+     * want to look at XCR0 bits and resolve the size in a new function similar to
+     * resolve_var_x64_size(). At the same time, we would want to change the mask register
+     * sizes in the decode table reflecting a simple variable size for all instructions.
```

Comment:
But isn't the mask size in any one instruction encoding fixed?  It depends on the total xmm/ymm/zmm register size and the element size, which are specified in the instruction.

The issue is that we're treating these like normal registers and having users say `opnd_create_reg(DR_REG_K1)` with no indication of size.  That ends up with `reg_get_size` of 64 bits.  If we wanted to get really accurate, we would have the INSTR_CREATE_ macros take in the raw reg enum instead of opnd for the mask regs, and have it supply the size particular to that instruction.

If we leave it how it is today, with all mask regs just a vague 64-bit size, I would still expect the code here to complain if I pass in a 128-bit register.  I don't see much point in checking `opsize`: that's *our* fixed size in the template and it never changes.  If we want some debug sanity checks on that size in our tables, let's put them elsewhere, not here on every encoding.

Reply:
Agree that this is basically self-checking as long as somebody is using the create macros. And yes, the mask size should be fixed in a given encoding.

>  If we leave it how it is today, with all mask regs just a vague 64-bit size, I would still expect the code here to complain if I pass in a 128-bit register.

But isn't this what it does? reg_size_ok checks for all the valid sizes that the different K types can have. I think it's best to take out the comment again, you confused me by not wanting this code here. But are you proposing to just return true for any TYPE_K_*?

Reply:
Updated, but I left in the opsize checks for TYPE_K_*. I'll follow up.

Reply:
> > If we leave it how it is today, with all mask regs just a vague 64-bit size, I would still expect the code here to complain if I pass in a 128-bit register.
>
> But isn't this what it does? reg_size_ok checks for all the valid sizes that the different K types can have

No, that is not what it does today.  The code today only looks at `opsize`, which is from the template.  It does not even look at the size of the register in the instruction being encoded.

Reply:
Are you proposing to remove reg_size_ok from opnd_type_ok for the types TYPE_K_MODRM_R, TYPE_K_VEX and TYPE_K_EVEX, because that's baked into the encoding anyway?

Reply:
Certainly the existing template size checks should not be there on every encoding, as already discussed.  I keep saying that it should actually check the register being considered (instead of ignoring the register and only looking at the template size.  I'm not sure how to make it clearer: check `reg` instead of `opsize`.  Given that mask registers have sub-sizes, I would say that it should use opnd_get_size() to support different size specifiers for mask registers.  Today we don't have the size set in the INSTR_CREATE_ macros, but maybe we should, as proposed above?  Even if we don't, a user might set them.

However, it looks like reg_size_ok() takes in just the reg_id_t, instead of the opnd, so it doesn't have the reg size.  So my proposed check is a bigger change than I thought and probably should not go into this PR.  At least add a comment about more fine-grained modeling of mask registers, and how they have particular sub-sizes yet today we neither set nor check any sizes.

Reply:
reg_size_ok() checks removed, and the following comment added:

```
* TODO i#1312: reg_size_ok() should consume the reg opnd, and validate its size.                                                                                                                          
* At the same time, the INSTR_CREATE_ macros should consume a size in addition to                                                                                                                         
* the mask register. Currently, mask register sizes are not checked properly and                                                                                                                          
* default to OPSZ_64.
```